Package request: john with opencl support

**Project description**

Is it possible to add OpenCL support to https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/security/john/default.nix ?

Details can be found by link below:
https://github.com/openwall/john/blob/bleeding-jumbo/doc/README-OPENCL


Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

I tried [here](https://github.com/NixOS/nixpkgs/compare/master...CyberShadow:pull-20240112-093826) but I'm getting a weird linker error:

```
/nix/store/fzlkaj1ax7gl655blfcr6zzvml1vx3bj-binutils-2.40/bin/ld: opencl_DES_bs_f_plug.o:/build/source/src/opencl_DES_bs.h:65: multiple definition of `opencl_DES_bs_index768'; opencl_DES_bs_b_plug.o:/build/source/src/opencl_DES_bs.h:65: first defined here
/nix/store/fzlkaj1ax7gl655blfcr6zzvml1vx3bj-binutils-2.40/bin/ld: opencl_DES_bs_h_plug.o:/build/source/src/opencl_DES_bs.h:65: multiple definition of `opencl_DES_bs_index768'; opencl_DES_bs_b_plug.o:/build/source/src/opencl_DES_bs.h:65: first defined here
/nix/store/fzlkaj1ax7gl655blfcr6zzvml1vx3bj-binutils-2.40/bin/ld: opencl_DES_bs_plug.o:/build/source/src/opencl_DES_bs.h:65: multiple definition of `opencl_DES_bs_index768'; opencl_DES_bs_b_plug.o:/build/source/src/opencl_DES_bs.h:65: first defined here
/nix/store/fzlkaj1ax7gl655blfcr6zzvml1vx3bj-binutils-2.40/bin/ld: opencl_DES_fmt_plug.o:/build/source/src/opencl_DES_bs.h:65: multiple definition of `opencl_DES_bs_index768'; opencl_DES_bs_b_plug.o:/build/source/src/opencl_DES_bs.h:65: first defined here
/nix/store/fzlkaj1ax7gl655blfcr6zzvml1vx3bj-binutils-2.40/bin/ld: opencl_axcrypt_fmt_plug.o:/build/source/src/opencl_axcrypt_fmt_plug.c:77: multiple definition of `insize'; opencl_agilekeychain_fmt_plug.o:/build/source/src/opencl_agilekeychain_fmt_plug.c:79: first defined here
/nix/store/fzlkaj1ax7gl655blfcr6zzvml1vx3bj-binutils-2.40/bin/ld: opencl_axcrypt_fmt_plug.o:/build/source/src/opencl_axcrypt_fmt_plug.c:77: multiple definition of `outsize'; opencl_agilekeychain_fmt_plug.o:/build/source/src/opencl_agilekeychain_fmt_plug.c:79: first defined here
/nix/store/fzlkaj1ax7gl655blfcr6zzvml1vx3bj-binutils-2.40/bin/ld: opencl_axcrypt_fmt_plug.o:/build/source/src/opencl_axcrypt_fmt_plug.c:77: multiple definition of `settingsize'; opencl_agilekeychain_fmt_plug.o:/build/source/src/opencl_agilekeychain_fmt_plug.c:79: first defined here
/nix/store/fzlkaj1ax7gl655blfcr6zzvml1vx3bj-binutils-2.40/bin/ld: opencl_axcrypt_fmt_plug.o:/build/source/src/opencl_axcrypt_fmt_plug.c:77: multiple definition of `cracked_size'; opencl_agilekeychain_fmt_plug.o:/build/source/src/opencl_agilekeychain_fmt_plug.c:79: first defined here
/nix/store/fzlkaj1ax7gl655blfcr6zzvml1vx3bj-binutils-2.40/bin/ld: opencl_gpg_fmt_plug.o:/build/source/src/opencl_gpg_fmt_plug.c:104: multiple definition of `settingsize'; opencl_agilekeychain_fmt_plug.o:/build/source/src/opencl_agilekeychain_fmt_plug.c:79: first defined here
...
```

I have no idea what causes it. Looking at the source, it looks perfectly fine.

--- a/pkgs/tools/security/john/default.nix
+++ b/pkgs/tools/security/john/default.nix
@@ -14,6 +14,10 @@
   python3Packages,
   perl,
   perlPackages,
+  withOpenCL ? true,
+  opencl-headers,
+  ocl-icd,
+  substituteAll,
   makeWrapper,
 }:
 
@@ -28,6 +32,13 @@ stdenv.mkDerivation rec {
     hash = "sha256-zvoN+8Sx6qpVg2JeRLOIH1ehfl3tFTv7r5wQZ44Qsbc=";
   };
 
+  patches = lib.optionals withOpenCL [
+    (substituteAll {
+      src = ./opencl.patch;
+      ocl_icd = ocl-icd;
+    })
+  ];
+
   postPatch = ''
     sed -ri -e '
       s!^(#define\s+CFG_[A-Z]+_NAME\s+).*/!\1"'"$out"'/etc/john/!
@@ -39,28 +50,37 @@ stdenv.mkDerivation rec {
     }' run/*.conf
   '';
 
-  preConfigure = ''
-    cd src
-    # Makefile.in depends on AS and LD being set to CC, which is set by default in configure.ac.
-    # This ensures we override the environment variables set in cc-wrapper/setup-hook.sh
-    export AS=$CC
-    export LD=$CC
-  '';
+  preConfigure =
+    ''
+      cd src
+      # Makefile.in depends on AS and LD being set to CC, which is set by default in configure.ac.
+      # This ensures we override the environment variables set in cc-wrapper/setup-hook.sh
+      export AS=$CC
+      export LD=$CC
+    ''
+    + lib.optionalString withOpenCL ''
+      python ./opencl_generate_dynamic_loader.py  # Update opencl_dynamic_loader.c
+    '';
   configureFlags = [
     "--disable-native-tests"
     "--with-systemwide"
   ];
 
-  buildInputs = [
-    openssl
-    nss
-    nspr
-    libkrb5
-    gmp
-    zlib
-    libpcap
-    re2
-  ];
+  buildInputs =
+    [
+      openssl
+      nss
+      nspr
+      libkrb5
+      gmp
+      zlib
+      libpcap
+      re2
+    ]
+    ++ lib.optionals withOpenCL [
+      opencl-headers
+      ocl-icd
+    ];
   nativeBuildInputs = [
     gcc
     python3Packages.wrapPython

--- a/pkgs/tools/security/john/default.nix
+++ b/pkgs/tools/security/john/default.nix
@@ -1,5 +1,21 @@
-{ lib, stdenv, fetchFromGitHub, openssl, nss, nspr, libkrb5, gmp, zlib, libpcap, re2
-, gcc, python3Packages, perl, perlPackages, makeWrapper, }:
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  openssl,
+  nss,
+  nspr,
+  libkrb5,
+  gmp,
+  zlib,
+  libpcap,
+  re2,
+  gcc,
+  python3Packages,
+  perl,
+  perlPackages,
+  makeWrapper,
+}:
 
 stdenv.mkDerivation rec {
   pname = "john";
@@ -35,13 +51,40 @@ stdenv.mkDerivation rec {
     "--with-systemwide"
   ];
 
-  buildInputs = [ openssl nss nspr libkrb5 gmp zlib libpcap re2 ];
-  nativeBuildInputs = [ gcc python3Packages.wrapPython perl makeWrapper ];
-  propagatedBuildInputs = (with python3Packages; [ dpkt scapy lxml ]) ++ # For pcap2john.py
-                          (with perlPackages; [ DigestMD4 DigestSHA1 GetoptLong # For pass_gen.pl
-                                                CompressRawLzma # For 7z2john.pl
-                                                perlldap ]); # For sha-dump.pl
-                          # TODO: Get dependencies for radius2john.pl and lion2john-alt.pl
+  buildInputs = [
+    openssl
+    nss
+    nspr
+    libkrb5
+    gmp
+    zlib
+    libpcap
+    re2
+  ];
+  nativeBuildInputs = [
+    gcc
+    python3Packages.wrapPython
+    perl
+    makeWrapper
+  ];
+  propagatedBuildInputs =
+    # For pcap2john.py
+    (with python3Packages; [
+      dpkt
+      scapy
+      lxml
+    ])
+    ++ (with perlPackages; [
+      # For pass_gen.pl
+      DigestMD4
+      DigestSHA1
+      GetoptLong
+      # For 7z2john.pl
+      CompressRawLzma
+      # For sha-dump.pl
+      perlldap
+    ]);
+  # TODO: Get dependencies for radius2john.pl and lion2john-alt.pl
 
   # gcc -DAC_BUILT -Wall vncpcap2john.o memdbg.o -g    -lpcap -fopenmp -o ../run/vncpcap2john
   # gcc: error: memdbg.o: No such file or directory
@@ -70,7 +113,11 @@ stdenv.mkDerivation rec {
     description = "John the Ripper password cracker";
     license = licenses.gpl2Plus;
     homepage = "https://github.com/openwall/john/";
-    maintainers = with maintainers; [ offline matthewbauer cherrykitten ];
+    maintainers = with maintainers; [
+      offline
+      matthewbauer
+      cherrykitten
+    ];
     platforms = platforms.unix;
   };
 }

--- a/pkgs/tools/security/john/opencl.patch
+++ b/pkgs/tools/security/john/opencl.patch
@@ -1,0 +1,18 @@
+diff --git a/src/opencl_generate_dynamic_loader.py b/src/opencl_generate_dynamic_loader.py
+index 790705330..2acedbc56 100755
+--- a/src/opencl_generate_dynamic_loader.py
++++ b/src/opencl_generate_dynamic_loader.py
+@@ -134,12 +134,7 @@ static void load_opencl_dll(void)
+ 
+ 	/* Names to try to load */
+ 	const char * const opencl_names[] = {
+-		"libOpenCL.so",		/* Linux/others, hack via "development" sub-package's symlink */
+-		"OpenCL",		/* _WIN */
+-		"/System/Library/Frameworks/OpenCL.framework/OpenCL", /* __APPLE__ */
+-		"opencl.dll",		/* __CYGWIN__ */
+-		"cygOpenCL-1.dll",	/* __CYGWIN__ */
+-		"libOpenCL.so.1"	/* Linux/others, no "development" sub-package installed */
++		"@ocl_icd@/lib/libOpenCL.so"	/* NixOS */
+ 	};
+ 
+ 	for (i = 0; i < sizeof(opencl_names)/sizeof(opencl_names[0]); i++) {

